primitives/message-lane/src/target_chain.rs

```diff
@@ -49,14 +61,35 @@ pub trait SourceHeaderChain<Fee> {
 	/// Error type.
 	type Error: Debug + Into<&'static str>;
 
-	/// Proof that messages are sent from source chain.
+	/// Proof that messages are sent from source chain. This may also include proof
+	/// of corresponding outbound lane states.
 	type MessagesProof: Parameter;
 
 	/// Verify messages proof and return proved messages.
 	///
 	/// Messages vector is required to be sorted by nonce within each lane. Out-of-order
 	/// messages will be rejected.
-	fn verify_messages_proof(proof: Self::MessagesProof) -> Result<Vec<Message<Fee>>, Self::Error>;
+	fn verify_messages_proof(proof: Self::MessagesProof) -> Result<ProvedMessages<Message<Fee>>, Self::Error>;
+
+	/// Verify messages proof and return proved messages with decoded payload.
+	fn verify_and_decode_messages_proof<DispatchPayload: Decode>(
+		proof: Self::MessagesProof,
+	) -> Result<ProvedMessages<DispatchMessage<DispatchPayload, Fee>>, Self::Error> {
+		Self::verify_messages_proof(proof).map(|messages_by_lane| {
+			messages_by_lane
+				.into_iter()
+				.map(|(lane, lane_data)| {
+					(
+						lane,
+						ProvedLaneMessages {
+							lane_state: lane_data.lane_state,
+							messages: lane_data.messages.into_iter().map(Into::into).collect(),
```

Comment:
We don't verify here that `messages` are actually coming from the same `lane`. While in theory it shouldn't happen, having this data duplicated may just lead to mistakes. Maybe let's change `Message/DispatchMessage` to only contain `nonce` (instead of entire `key`) and add the lane information only during actual `receive_message`?

Reply:
> We don't verify here that messages are actually coming from the same lane. While in theory it shouldn't happen, having this data duplicated may just lead to mistakes. Maybe let's change Message/DispatchMessage to only contain nonce (instead of entire key) and add the lane information only during actual receive_message?

`verify_messages_proof` could return anything it wants - it can e.g. replace all messages payloads with `BalancesCall::transfer(svyatonik, 1 DOT)`. So you just need to trust it.

We can't remove `lane` from the MessageKey, because in this way relayer can just submit messages from Lane1 to Lane2.

Reply:
My point was that this information is now duplicated. We already have lane id in the returned map, why do we need it in the message itself?

Reply:
Ok - will remove from trait

Reply:
done